presto-hive/src/main/java/com/facebook/presto/hive/util/MergingPageIterator.java

```diff
@@ -0,0 +1,142 @@
+import com.google.common.collect.AbstractIterator;
+import com.google.common.collect.ImmutableList;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+
+import static com.facebook.presto.hive.util.TempFileWriter.appendPositionTo;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.Iterators.concat;
+import static com.google.common.collect.Iterators.mergeSorted;
+import static com.google.common.collect.Iterators.transform;
+import static java.util.Comparator.naturalOrder;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+
+public class MergingPageIterator
+        extends AbstractIterator<Page>
+{
+    private final List<Type> types;
+    private final List<Integer> sortFields;
+    private final List<SortOrder> sortOrders;
+    private final PageBuilder pageBuilder;
+    private final Iterator<PagePosition> pagePositions;
+
+    public MergingPageIterator(
+            Collection<Iterator<Page>> iterators,
+            List<Type> types,
+            List<Integer> sortFields,
+            List<SortOrder> sortOrders)
+    {
+        requireNonNull(sortFields, "sortFields is null");
+        requireNonNull(sortOrders, "sortOrders is null");
+        checkArgument(sortFields.size() == sortOrders.size(), "sortFields and sortOrders size must match");
+
+        this.types = ImmutableList.copyOf(requireNonNull(types, "types is null"));
+        this.sortFields = ImmutableList.copyOf(sortFields);
+        this.sortOrders = ImmutableList.copyOf(sortOrders);
+        this.pageBuilder = new PageBuilder(types);
+        this.pagePositions = mergeSorted(
+                iterators.stream()
+                        .map(pages -> concat(transform(pages, PagePositionIterator::new)))
+                        .collect(toList()),
+                naturalOrder());
+    }
+
+    @Override
+    protected Page computeNext()
+    {
+        while (!pageBuilder.isFull() && pagePositions.hasNext()) {
+            pagePositions.next().appendTo(pageBuilder);
+        }
+
+        if (pageBuilder.isEmpty()) {
+            return endOfData();
+        }
+
+        Page page = pageBuilder.build();
+        pageBuilder.reset();
+        return page;
+    }
+
+    private class PagePositionIterator
+            extends AbstractIterator<PagePosition>
+    {
+        private final Page page;
+        private int position = -1;
+
+        private PagePositionIterator(Page page)
+        {
+            this.page = requireNonNull(page, "page is null");
+        }
+
+        @Override
+        protected PagePosition computeNext()
+        {
+            position++;
+            if (position == page.getPositionCount()) {
+                return endOfData();
+            }
+            return new PagePosition(page, position);
+        }
+    }
+
+    @SuppressWarnings("ComparableImplementedButEqualsNotOverridden")
+    private class PagePosition
```

Comment:
We should consider exposing a method to get a PagesIndexComparator

Reply:
That takes a `PagesIndex` is a full index of a single set of pages. This is merging multiple sets of sorted pages.